src/Data/LLVM/BitCode/GetBits.hs

```diff
@@ -11,68 +11,106 @@ module Data.LLVM.BitCode.GetBits (
   , skip
   ) where
 
-import Data.LLVM.BitCode.BitString
+import           Data.LLVM.BitCode.BitString
 
-import Control.Applicative (Alternative(..))
-import Control.Arrow (first)
-import Control.Monad (MonadPlus(..),when,replicateM_)
-import qualified Data.Binary.Get as BG
-import Data.Bits (shiftR)
-import Data.ByteString (ByteString)
-import Data.Word (Word32)
+import           Control.Applicative ( Alternative(..) )
+import           Control.Monad ( MonadPlus(..) )
+import           Data.Bits ( (.&.), (.|.), shiftL, shiftR )
+import           Data.ByteString ( ByteString )
+import qualified Data.ByteString as BS
 
 #if !MIN_VERSION_base(4,13,0)
-import Control.Monad.Fail( MonadFail )
+import           Control.Monad.Fail ( MonadFail )
 import qualified Control.Monad.Fail
 #endif
 
 -- Bit-level Parsing -----------------------------------------------------------
 
-newtype GetBits a = GetBits { unGetBits :: SubWord -> BG.Get (a,SubWord) }
+newtype GetBits a = GetBits { unGetBits :: Either String (BitPosition -> BS.ByteString -> (BitsGetter a, BitPosition)) } -- Left is outer fail
+
+newtype BitPosition = BitPosition (NumBits, NumBits) -- (current, limit)
+
+type BitsGetter a = BS.ByteString -> Either String a -- Left is inner fail
+
 
 -- | Run a @GetBits@ action, returning its value, and the number of bits offset
 -- into the next byte of the stream.
-runGetBits :: GetBits a -> BG.Get a
-runGetBits m = fst `fmap` unGetBits m aligned
+runGetBits :: GetBits a -> ByteString -> Either String a
+runGetBits m bs =
+  let startPos = BitPosition ( Bits' 0
+                             , bytesToBits $ Bytes' $ BS.length bs
+                             )
+  in ($ bs) =<< (fst . ($ bs) . ($ startPos) <$> unGetBits m)
```

Comment:
I have a really hard time reading this code due to the excessive use of pointfree code.

Reply:
Noted.  This particular line will change in an upcoming PR, but I'll also keep the general concern in mind going forward.